browser/ui/webui/padded_image_source.cc

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2025 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/browser/ui/webui/padded_image_source.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "brave/brave_domains/service_domains.h"
+#include "brave/components/brave_private_cdn/private_cdn_helper.h"
+#include "chrome/browser/profiles/profile.h"
+#include "net/base/net_errors.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+
+PaddedImageSource::PaddedImageSource(Profile* profile)
+    : SanitizedImageSource(profile),
+      pcdn_domain_(brave_domains::GetServicesDomain("pcdn")) {}
+
+PaddedImageSource::PaddedImageSource(
+    Profile* profile,
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory,
+    std::unique_ptr<DataDecoderDelegate> delegate,
+    std::string pcdn_domain)
+    : SanitizedImageSource(profile, url_loader_factory, std::move(delegate)),
+      pcdn_domain_(pcdn_domain) {}
+
+PaddedImageSource::~PaddedImageSource() = default;
+
+void PaddedImageSource::OnImageLoaded(
+    std::unique_ptr<network::SimpleURLLoader> loader,
+    RequestAttributes request_attributes,
+    content::URLDataSource::GotDataCallback callback,
+    std::unique_ptr<std::string> body) {
+  if (loader->NetError() != net::OK || !body) {
+    std::move(callback).Run(nullptr);
+    return;
+  }
+
+  if (request_attributes.image_url.host() == pcdn_domain_ &&
+      request_attributes.image_url.path().ends_with(".pad")) {
+    std::string_view body_payload(body->data(), body->size());
+    if (!brave::PrivateCdnHelper::GetInstance()->RemovePadding(&body_payload)) {
+      std::move(callback).Run(nullptr);
+      return;
+    }
+
+    // TODO: Ask a C++ person if this is safe. |body_payload| points to a
+    // substring of |body|, so I'm not sure if we need to force a copy first.
+    *body = body_payload;
```

Comment:
@cdesouza-chromium is this something I can do, or am I shooting myself in the foot?

Reply:
Resolving - talked to @cdesouza-chromium and he says its not an issue :smile:

Reply:
It's fine.